File: results/1.0/MQTT/mqtt-claude-3-5-sonnet-20241022.c
#include <hammer/hammer.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Forward declarations
HParser* init_mqtt_parser(void);
HParser* mqtt_fixed_header(void);
HParser* mqtt_variable_header(void);
HParser* mqtt_payload(void);
HParser* mqtt_properties(void);

// Property Parsers
HParser* property_single_byte(void) {
    return h_uint8();
}

HParser* property_two_byte(void) {
    return h_uint16();
}

HParser* property_four_byte(void) {
    return h_uint32();
}

HParser* property_var_int(void) {
    return h_many(h_uint8());
}

HParser* property_utf8_string(void) {
    return h_sequence(h_uint16(), h_many(h_uint8()), NULL);
}

HParser* property_binary_data(void) {
    return h_sequence(h_uint16(), h_many(h_uint8()), NULL);
}

HParser* mqtt_properties(void) {
    return h_sequence(
        h_uint8(),  // Property Length
        h_many(h_choice(
            h_sequence(h_ch(0x01), property_single_byte(), NULL),    // Payload Format Indicator
            h_sequence(h_ch(0x02), property_four_byte(), NULL),      // Message Expiry Interval
            h_sequence(h_ch(0x03), property_utf8_string(), NULL),    // Content Type
            h_sequence(h_ch(0x08), property_utf8_string(), NULL),    // Response Topic
            h_sequence(h_ch(0x09), property_binary_data(), NULL),    // Correlation Data
            h_sequence(h_ch(0x0B), property_var_int(), NULL),        // Subscription Identifier
            h_sequence(h_ch(0x11), property_four_byte(), NULL),      // Session Expiry Interval
            h_sequence(h_ch(0x12), property_utf8_string(), NULL),    // Assigned Client Identifier
            h_sequence(h_ch(0x13), property_two_byte(), NULL),       // Server Keep Alive
            h_sequence(h_ch(0x15), property_utf8_string(), NULL),    // Authentication Method
            h_sequence(h_ch(0x16), property_binary_data(), NULL),    // Authentication Data
            h_sequence(h_ch(0x17), property_single_byte(), NULL),    // Request Problem Information
            h_sequence(h_ch(0x19), property_single_byte(), NULL),    // Request Response Information
            h_sequence(h_ch(0x1A), property_utf8_string(), NULL),    // Response Information
            h_sequence(h_ch(0x1C), property_utf8_string(), NULL),    // Server Reference
            h_sequence(h_ch(0x1F), property_utf8_string(), NULL),    // Reason String
            h_sequence(h_ch(0x21), property_two_byte(), NULL),       // Receive Maximum
            h_sequence(h_ch(0x22), property_two_byte(), NULL),       // Topic Alias Maximum
            h_sequence(h_ch(0x23), property_two_byte(), NULL),       // Topic Alias
            h_sequence(h_ch(0x24), property_single_byte(), NULL),    // Maximum QoS
            h_sequence(h_ch(0x25), property_single_byte(), NULL),    // Retain Available
            h_sequence(h_ch(0x26), property_utf8_string(), NULL),    // User Property
            h_sequence(h_ch(0x27), property_four_byte(), NULL),      // Maximum Packet Size
            h_sequence(h_ch(0x28), property_single_byte(), NULL),    // Wildcard Subscription Available
            h_sequence(h_ch(0x29), property_single_byte(), NULL),    // Subscription Identifier Available
            h_sequence(h_ch(0x2A), property_single_byte(), NULL),    // Shared Subscription Available
            NULL)),
        NULL);
}

HParser* mqtt_fixed_header(void) {
    return h_sequence(
        h_bits(4, false),  // Packet Type
        h_bits(4, false),  // Flags
        h_many1(h_uint8()), // Remaining Length
        NULL);
}

HParser* mqtt_connect_variable_header(void) {
    return h_sequence(
        h_uint16(),        // Protocol Name Length
        h_many(h_uint8()), // Protocol Name
        h_uint8(),         // Protocol Version
        h_bits(8, false),  // Connect Flags
        h_uint16(),        // Keep Alive
        mqtt_properties(),
        NULL);
}

HParser* mqtt_connect_payload(void) {
    return h_sequence(
        h_uint16(),        // Client ID Length
        h_many(h_uint8()), // Client ID
        h_optional(h_sequence(
            mqtt_properties(),     // Will Properties
            h_uint16(),           // Will Topic Length
            h_many(h_uint8()),    // Will Topic
            h_uint16(),           // Will Payload Length
            h_many(h_uint8()),    // Will Payload
            NULL)),
        h_optional(h_sequence(
            h_uint16(),           // Username Length
            h_many(h_uint8()),    // Username
            NULL)),
        h_optional(h_sequence(
            h_uint16(),           // Password Length
            h_many(h_uint8()),    // Password
            NULL)),
        NULL);
}

HParser* mqtt_publish_variable_header(void) {
    return h_sequence(
        h_uint16(),        // Topic Name Length
        h_many(h_uint8()), // Topic Name
        h_optional(h_uint16()), // Packet Identifier (if QoS > 0)
        mqtt_properties(),
        NULL);
}

HParser* mqtt_publish_payload(void) {
    return h_many(h_uint8()); // Application Message
}

HParser* mqtt_subscribe_payload(void) {
    return h_many1(h_sequence(
        h_uint16(),        // Topic Filter Length
        h_many(h_uint8()), // Topic Filter
        h_uint8(),         // Subscription Options
        NULL));
}

HParser* init_mqtt_parser(void) {
    return h_sequence(
        mqtt_fixed_header(),
        h_choice(
            h_sequence(mqtt_connect_variable_header(), mqtt_connect_payload(), NULL),
            h_sequence(mqtt_publish_variable_header(), mqtt_publish_payload(), NULL),
            h_sequence(h_uint16(), mqtt_properties(), NULL), // For PUBACK, PUBREC, PUBREL, PUBCOMP
            h_sequence(h_uint16(), mqtt_properties(), mqtt_subscribe_payload(), NULL), // SUBSCRIBE
            h_sequence(h_uint16(), mqtt_properties(), NULL), // SUBACK
            h_sequence(h_uint16(), mqtt_properties(), h_many(h_uint8()), NULL), // UNSUBSCRIBE
            h_sequence(h_uint16(), mqtt_properties(), NULL), // UNSUBACK
            h_sequence(h_uint8(), mqtt_properties(), NULL), // DISCONNECT
            h_sequence(h_uint8(), mqtt_properties(), NULL), // AUTH
            NULL),
        NULL);
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <input_file>\n", argv[0]);
        return 1;
    }

    FILE *file = fopen(argv[1], "rb");
    if (!file) {
        perror("Failed to open file");
        return 1;
    }

    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    uint8_t *input = malloc(file_size);
    if (!input) {
        perror("Failed to allocate memory");
        fclose(file);
        return 1;
    }

    size_t bytes_read = fread(input, 1, file_size, file);
    fclose(file);

    if (bytes_read != file_size) {
        fprintf(stderr, "Failed to read entire file\n");
        free(input);
        return 1;
    }

    HParser *mqtt_parser = init_mqtt_parser();
    if (!mqtt_parser) {
        fprintf(stderr, "Failed to initialize parser\n");
        free(input);
        return 1;
    }

    HParseResult *result = h_parse(mqtt_parser, input, file_size);
    if (!result) {
        fprintf(stderr, "Parse failed\n");
        free(input);
        return 1;
    }

    // Success
    fprintf(stdout, "Parse successful\n");
    
    h_parse_result_free(result);
    free(input);
    return 0;
}

File: results/1.0/MQTT/mqtt-claude-3-5-sonnet-20241022.rs
use nom::{
    bits::complete::{tag as bits_tag, take as bits_take},
    bytes::complete::{tag, take},
    combinator::{map, verify},
    multi::{length_data, many0},
    number::complete::{be_u16, be_u32, be_u8},
    sequence::tuple,
    IResult,
};
use std::{env, fs, io::Read};

#[derive(Debug)]
enum PacketType {
    CONNECT = 1,
    CONNACK = 2,
    PUBLISH = 3,
    PUBACK = 4,
    PUBREC = 5,
    PUBREL = 6,
    PUBCOMP = 7,
    SUBSCRIBE = 8,
    SUBACK = 9,
    UNSUBSCRIBE = 10,
    UNSUBACK = 11,
    PINGREQ = 12,
    PINGRESP = 13,
    DISCONNECT = 14,
    AUTH = 15,
}

#[derive(Debug)]
struct FixedHeader {
    packet_type: PacketType,
    dup_flag: bool,
    qos_level: u8,
    retain: bool,
    remaining_length: u32,
}

#[derive(Debug)]
struct Property {
    identifier: u8,
    value: Vec<u8>,
}

#[derive(Debug)]
struct ConnectFlags {
    username_flag: bool,
    password_flag: bool,
    will_retain: bool,
    will_qos: u8,
    will_flag: bool,
    clean_start: bool,
}

#[derive(Debug)]
struct ConnectPacket {
    protocol_name: String,
    protocol_version: u8,
    connect_flags: ConnectFlags,
    keep_alive: u16,
    properties: Vec<Property>,
    client_id: String,
    will_properties: Option<Vec<Property>>,
    will_topic: Option<String>,
    will_payload: Option<Vec<u8>>,
    username: Option<String>,
    password: Option<Vec<u8>>,
}

fn parse_remaining_length(input: &[u8]) -> IResult<&[u8], u32> {
    let mut multiplier: u32 = 1;
    let mut value: u32 = 0;
    let mut current_input = input;
    let mut byte_count = 0;

    loop {
        let (rest, byte) = be_u8(current_input)?;
        current_input = rest;
        byte_count += 1;

        value += ((byte & 0x7F) as u32) * multiplier;
        multiplier *= 128;

        if byte & 0x80 == 0 || byte_count >= 4 {
            break;
        }
    }

    Ok((current_input, value))
}

fn parse_fixed_header(input: &[u8]) -> IResult<&[u8], FixedHeader> {
    let (input, (packet_type_bits, flags)) = bits::bits::<_, _, Error<(&[u8], usize)>, _, _>(
        tuple((
            bits_take(4usize),
            bits_take(4usize),
        ))
    )(input)?;

    let packet_type = match packet_type_bits {
        1 => PacketType::CONNECT,
        2 => PacketType::CONNACK,
        3 => PacketType::PUBLISH,
        4 => PacketType::PUBACK,
        5 => PacketType::PUBREC,
        6 => PacketType::PUBREL,
        7 => PacketType::PUBCOMP,
        8 => PacketType::SUBSCRIBE,
        9 => PacketType::SUBACK,
        10 => PacketType::UNSUBSCRIBE,
        11 => PacketType::UNSUBACK,
        12 => PacketType::PINGREQ,
        13 => PacketType::PINGRESP,
        14 => PacketType::DISCONNECT,
        15 => PacketType::AUTH,
        _ => return Err(nom::Err::Error(Error::new(input, ErrorKind::Tag))),
    };

    let (input, remaining_length) = parse_remaining_length(input)?;

    Ok((input, FixedHeader {
        packet_type,
        dup_flag: (flags & 0x08) != 0,
        qos_level: (flags & 0x06) >> 1,
        retain: (flags & 0x01) != 0,
        remaining_length,
    }))
}

fn parse_utf8_string(input: &[u8]) -> IResult<&[u8], String> {
    let (input, len) = be_u16(input)?;
    let (input, string_bytes) = take(len)(input)?;
    Ok((input, String::from_utf8_lossy(string_bytes).into_owned()))
}

fn parse_property(input: &[u8]) -> IResult<&[u8], Property> {
    let (input, identifier) = be_u8(input)?;
    let (input, value) = match identifier {
        0x01 | 0x17 | 0x24 | 0x25 | 0x28 | 0x29 | 0x2A => {
            let (input, val) = be_u8(input)?;
            (input, vec![val])
        },
        0x02 | 0x11 | 0x21 | 0x27 => {
            let (input, val) = be_u32(input)?;
            (input, val.to_be_bytes().to_vec())
        },
        0x22 | 0x23 => {
            let (input, val) = be_u16(input)?;
            (input, val.to_be_bytes().to_vec())
        },
        _ => {
            let (input, str_val) = parse_utf8_string(input)?;
            (input, str_val.into_bytes())
        },
    };
    
    Ok((input, Property { identifier, value }))
}

fn parse_properties(input: &[u8]) -> IResult<&[u8], Vec<Property>> {
    let (input, length) = parse_remaining_length(input)?;
    let (input, props) = many0(parse_property)(input)?;
    Ok((input, props))
}

fn parse_connect_flags(input: &[u8]) -> IResult<&[u8], ConnectFlags> {
    let (input, flags) = be_u8(input)?;
    Ok((input, ConnectFlags {
        username_flag: (flags & 0x80) != 0,
        password_flag: (flags & 0x40) != 0,
        will_retain: (flags & 0x20) != 0,
        will_qos: (flags & 0x18) >> 3,
        will_flag: (flags & 0x04) != 0,
        clean_start: (flags & 0x02) != 0,
    }))
}

fn parse_connect(input: &[u8]) -> IResult<&[u8], ConnectPacket> {
    let (input, protocol_name) = parse_utf8_string(input)?;
    let (input, protocol_version) = verify(be_u8, |&x| x == 5)(input)?;
    let (input, connect_flags) = parse_connect_flags(input)?;
    let (input, keep_alive) = be_u16(input)?;
    let (input, properties) = parse_properties(input)?;
    let (input, client_id) = parse_utf8_string(input)?;

    let (input, will_properties, will_topic, will_payload) = if connect_flags.will_flag {
        let (input, props) = parse_properties(input)?;
        let (input, topic) = parse_utf8_string(input)?;
        let (input, payload) = length_data(be_u16)(input)?;
        (input, Some(props), Some(topic), Some(payload.to_vec()))
    } else {
        (input, None, None, None)
    };

    let (input, username) = if connect_flags.username_flag {
        let (input, name) = parse_utf8_string(input)?;
        (input, Some(name))
    } else {
        (input, None)
    };

    let (input, password) = if connect_flags.password_flag {
        let (input, pass) = length_data(be_u16)(input)?;
        (input, Some(pass.to_vec()))
    } else {
        (input, None)
    };

    Ok((input, ConnectPacket {
        protocol_name,
        protocol_version,
        connect_flags,
        keep_alive,
        properties,
        client_id,
        will_properties,
        will_topic,
        will_payload,
        username,
        password,
    }))
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        println!("Usage: {} <mqtt_binary_file>", args[0]);
        return Ok(());
    }

    let mut file = fs::File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_fixed_header(&buffer) {
        Ok((remaining, header)) => {
            println!("Fixed Header: {:?}", header);
            match header.packet_type {
                PacketType::CONNECT => {
                    match parse_connect(remaining) {
                        Ok((_, packet)) => println!("Connect Packet: {:?}", packet),
                        Err(e) => println!("Error parsing CONNECT: {:?}", e),
                    }
                },
                _ => println!("Parsing for packet type {:?} not implemented", header.packet_type),
            }
        },
        Err(e) => println!("Error parsing fixed header: {:?}", e),
    }

    Ok(())
}